beartype/_util/cache/pool/utilcachepool.py

```diff
@@ -189,6 +171,9 @@ def acquire(self, key: 'HashableType' = None) -> object:
                 else pool.pop()
             )
 
+            self._object_ids_acquired[id(pool_item)] = True
```

Comment:
>Excellent. I am stroking my wrinkled hands with glee.

I think you need moisturizer 😆  

> As a sanity check, we might also consider validating that this boolean isn't already `True` with a conditional resembling the following immediately _before_ the above assignment:
> 
> ```
>         if self._object_ids_acquired.get(id(item)):
>             raise _BeartypeUtilKeyPoolException(
>                 f'Key pool unreleased object {repr(item)} already acquired.')
> ```
> 
> ...or _something_ like that. Since the whole point of this refactoring exercise is to improve robustness, we'd might as well go the whole nine yards and punt for a risky field goal. What do you think?

Hmmm, based on the description in #13, that seems logical: if we're wanting to guard against attempting to release an unacquired item then conversely we'd want to guard against attempting to obtain an already acquired item. 

I'm wondering if it's strictly necessary though. The [official doc](https://docs.python.org/3.8/library/functions.html#id) says the result of `id()` is 

> "an integer which is guaranteed to be unique and constant for this object during its lifetime. Two objects with non-overlapping lifetimes may have the same id() value"

If the integer is protected while the Object is alive and freed upon it's death, wouldn't the only check we need to perform be on `release()`? 
Reasoning: Unlike `release()`, only a key is provided to `acquire()` - the system is in charge of either popping an existing object from FixedList or spinning a new object up. In either case, there should never be an instance where there's a conflict. Unless, and this **might** be crazy, there' a way to have a factory intentionally act as a bad actor using .

...Unless I've missed the point of a sanity check 🙈. Am I off in left field with this thought?

Reply:
> Am I off in left field with this thought?

Absolutely... not! You're correct, as usual. This means you're now in right field, where only the most die-hard of computational outfielders dear to tread. :baseball: 

> Unless, and this might be crazy, there' a way to have a factory intentionally act as a bad actor.

**That's it.** Since callers define the factory and the objects produced by that factory are otherwise *not* validated, it's feasible for a factory to erroneously produce duplicate objects (e.g., due to improper caching or memoization). You are now thinking: "@leycec, are you always this paranoid or are you merely daft?"

The answer to both questions is "Yes." Still, object duplication happens surprisingly often in Python – especially for objects that *internally* cache themselves, like strings and PEP 484-compliant type hints: e.g.,

```
>>> 'is' is 'is'
True
>>> import typing
>>> typing.List[int] is typing.List[int]
True
```

You are now thinking: "@leycec, why would objects that internally cache themselves need to be explicitly cached in a `KeyPool`, anyway?"

The answer to that question is, of course... *they don't.* Look, kid. My paranoia is only equalled by my daftness. This is what happens when you live in a cabin in the woods several miles from the nearest Royal Canadian Mounted Police (RCMP) station. Yes, we still have RCMP. They just ride steel horses now. Cue [gratuitous 80's song complete with crowd lighters during the acoustic intro](https://www.youtube.com/watch?v=SRvCvsRp5ho).

I probably should have justified my initial snippet with that rational. Alas, I didn't. Now, we all pay the price.